Add Ceph RGW S3 compatibility

jmunhoz opened the pull request:
##### Issue Type:
- Feature Pull Request (minimal impact in the current code)
##### Plugin Name:

./cloud/amazon/s3.py
##### Ansible Version:

```
$ ansible --version
ansible 2.1.0 (devel 3ddcabee0e) last updated 2016/02/11 11:21:11 (GMT +200)
  lib/ansible/modules/core: (s3-ceph 1da965e039) last updated 2016/02/11 14:27:39 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD 6aeb2ab6cf) last updated 2016/02/11 11:34:36 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides

```
##### Summary:

Ansible supports S3 on AWS, Walrus and FakeS3 but it doesn't support Ceph RGW S3. This patch supports Ceph RGW S3.

More information at:

https://groups.google.com/forum/#!topic/ansible-devel/XxVeoHehpSM
##### Example output:

```
```

Ceph Object Gateway (Ceph RGW) is an object storage interface built on top of
librados to provide applications with a RESTful gateway to Ceph Storage
Clusters:

http://docs.ceph.com/docs/master/radosgw/

This patch adds the required bits to use the RGW S3 RESTful API properly.

Signed-off-by: Javier M. Mellid jmunhoz@igalia.com

thereallukl commented:
How about adding extra is_rgw method instead of new parameter (https://github.com/lleszczu/ansible-modules-core/commit/810cd9c2143777831a1d5fb09e48bfb8d7525e2f)? That way it's consistent with fakes3.

jmunhoz commented:
@lleszczu I would prefer being consistent with Walrus and AWS instead of fakes3. fakes3 scheme encoding/decoding looks a quick way to add/test with fakes3. It would be weird having walrus and aws using http/https and rgw using rgw/rgws.

Anyway, it would be trivial to adapt the patch if it make sense at the end. I guess you miss the mail I share on ansible-devel [1], as suggested by @bcoca, 12 days ago. Maybe you could reply that mail with this approach in order to boost the discussion. What do you think?

[1] https://groups.google.com/forum/#!topic/ansible-devel/XxVeoHehpSM

Thanks for commenting!

thereallukl commented:
>  would prefer being consistent with Walrus and AWS instead of fakes3. fakes3 scheme encoding/decoding looks a quick way to add/test with fakes3

Seems reasonable
+1

jmunhoz committed:
Add Ceph RGW S3 compatibility

Ceph Object Gateway (Ceph RGW) is an object storage interface built on top of
librados to provide applications with a RESTful gateway to Ceph Storage
Clusters:

http://docs.ceph.com/docs/master/radosgw/

This patch adds the required bits to use the RGW S3 RESTful API properly.

Signed-off-by: Javier M. Mellid <jmunhoz@igalia.com>

robynbergeron commented:
Thanks @jmunhoz for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

robynbergeron commented:
Hey, @leseb -- could you maybe eyeball this from the ceph angle and weigh in? :)

leseb commented:
@robynbergeron sure, will do :)

leseb commented:
From a Ceph angle, I don't see any issue, so :+1: for me.

abadger commented on cloud/amazon/s3.py:
```diff
@@ -131,9 +131,13 @@
+  rgw:
+    description:
+      - Enable Ceph RGW S3 support
+    default: false
```
Needs to specify version_added: 2.2

abadger replied:
From bcoca: If this option is specific to ceph then there needs to be an indication in the docs that it is ignored when used against other S3 compatible endpoints.

abadger commented on cloud/amazon/s3.py:
```diff
@@ -590,13 +608,21 @@ def main():
+            try:
+                all_buckets = s3.get_all_buckets()
+                s3_keys = [x.name for x in all_buckets]
+            except s3.provider.storage_response_error, e:
```
Need to make sure that these compile under python3 (we'll shoot for actually working under python3 later).  It looks like boto is a requirement here and boto requires python>=2.6 so we can just do:

``` python
except s3.provider.storage_response_error as e:
```

catching exceptions in files that do not have libraries which require python >= 2.6 would use something like this instead:

``` python
from ansible.module_utils.py24compat import get_exception
[...]
except s3.provider.storage_response_error:
    e = get_exception()
```

abadger commented:
made a few general comments.  Our main aws guy is on vacation right now so looking around for someone else to review this change from an S3 angle.

ryansb commented:
Other than Toshio's syntax comments, the code itself looks good here. I don't have a ceph cluster to test, but the S3 bucket functionality isn't affected by this change.

jmunhoz committed:
Add proper version_added for rgw option in s3.py

Signed-off-by: Javier M. Mellid <jmunhoz@igalia.com>

jmunhoz committed:
Add s3_url requirement in doc when rgw support is enabled in s3.py

Signed-off-by: Javier M. Mellid <jmunhoz@igalia.com>

jmunhoz commented:
@abadger @ryansb Thanks for the reviewing. It is very appreciated. I added the mentioned changes on top of the original patches. Rebased to catch the head and tested with Ceph RGW. It works ok on my side.

ryansb commented on cloud/amazon/s3.py:
```diff
@@ -591,13 +610,21 @@ def main():
-            module.fail_json(msg="Target bucket (%s) cannot be found"% bucket, failed=True)
-
-        list_keys(module, bucket_object, prefix, marker, max_keys)
+        if bucket == '/':
```
For clarity, can you make this only happen on RGW flavors? This code is specific to RGW, and I think it convolutes what we had before since there's no (apparent) reason for special-casing buckets named `/`.

jmunhoz replied:
@ryansb I just removed the owned buckets listing operation instead of making it happen on RGW only. I think it could add unnecessary divergence against the AWS S3 interface in this moment.

jmunhoz commented:
@ryansb Pushed updated patchset tested with Ceph RGW.

ryansb commented:
LGTM
